akka-sample-fsm-scala/src/main/scala/sample/typed/DiningHakkersTyped.scala

```diff
@@ -0,0 +1,190 @@
+package sample.typed
+
+import akka.NotUsed
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.typed.{ActorRef, ActorSystem, Behavior}
+
+import scala.concurrent.duration._
+
+// Akka adaptation of
+// http://www.dalnefre.com/wp/2010/08/dining-philosophers-in-humus/
+
+/*
+* Some messages for the chopstick
+*/
+sealed trait ChopstickMessage
+final case class Take(ref: ActorRef[ChopstickAnswer]) extends ChopstickMessage
+final case class Put(ref: ActorRef[ChopstickAnswer]) extends ChopstickMessage
+
+sealed trait ChopstickAnswer
+final case class Taken(chopstick: ActorRef[ChopstickMessage]) extends ChopstickAnswer
+final case class Busy(chopstick: ActorRef[ChopstickMessage]) extends ChopstickAnswer
+
+/*
+* A Chopstick is an actor, it can be taken, and put back
+*/
+object Chopstick {
+
+  //When a Chopstick is taken by a hakker
+  //It will refuse to be taken by other hakkers
+  //But the owning hakker can put it back
+  def takenBy(hakker: ActorRef[ChopstickAnswer]): Behavior[ChopstickMessage] = {
+    Behaviors.receivePartial {
+      case (ctx, Take(otherHakker)) =>
+        otherHakker ! Busy(ctx.self)
+        Behaviors.same
+      case (ctx, Put(`hakker`)) =>
+        available
+    }
+  }
+
+  //When a Chopstick is available, it can be taken by a hakker
+  lazy val available: Behavior[ChopstickMessage] = {
+    Behaviors.receivePartial {
+      case (ctx, Take(hakker)) =>
+        hakker ! Taken(ctx.self)
+        takenBy(hakker)
+    }
+  }
+}
+
+/**
+  * Some fsm hakker messages
+  */
+sealed trait HakkerMessage
+case object Think extends HakkerMessage
+case object Eat extends HakkerMessage
+
+/*
+* A hakker is an awesome dude or dudette who either thinks about hacking or has to eat ;-)
+*/
+class Hakker(name: String, left: ActorRef[ChopstickMessage], right: ActorRef[ChopstickMessage]) {
+
+  final case class ChopstickAnswerAdaptor(msg: ChopstickAnswer) extends HakkerMessage
+
+  lazy val waiting: Behavior[HakkerMessage] =
+    Behaviors.receivePartial {
+      case (ctx, Think) =>
+        println(s"$name starts to think")
+        startThinking(ctx, 5.seconds)
+    }
+
+  //When a hakker is thinking it can become hungry
+  //and try to pick up its chopsticks and eat
+  lazy val thinking: Behavior[HakkerMessage] =
+    Behaviors.setup { ctx =>
+      val adapter = ctx.messageAdapter(ChopstickAnswerAdaptor)
+      Behaviors.receiveMessagePartial {
+        case Eat =>
+          left ! Take(adapter)
+          right ! Take(adapter)
+          hungry
+      }
```

Comment:
While `receive...Partial` is neat, I think I'd preferred explicit case with `unhandled` in a sample, to make things clear, or at least a comment noting that we use it to handle only a subset of the messages in each state.

Reply:
Done